xen/drivers/passthrough/arm/io-pgtable-arm.c

```diff
@@ -259,6 +312,40 @@ static void __arm_lpae_set_pte(arm_lpae_iopte *ptep, arm_lpae_iopte pte,
 					   __arm_lpae_dma_addr(ptep),
 					   sizeof(pte), DMA_TO_DEVICE);
 }
+#endif
+
+static struct page_info *__arm_lpae_alloc_pages(size_t size, gfp_t gfp,
+				    struct io_pgtable_cfg *cfg)
+{
+	struct page_info *pages;
+	unsigned int order = get_order_from_bytes(size);
+	int i;
+
+	pages = alloc_domheap_pages(NULL, order, 0);
+	if (pages == NULL)
+		return NULL;
+
+	for (i = 0; i < (1<<order); i ++)
+		clear_and_clean_page(pages + i);
+
+	return pages;
+}
+
+static void __arm_lpae_free_pages(struct page_info *pages, size_t size,
+				  struct io_pgtable_cfg *cfg)
+{
+	unsigned int order = get_order_from_bytes(size);
+
+	free_domheap_pages(pages, order);
+}
+
+static void __arm_lpae_set_pte(arm_lpae_iopte *ptep, arm_lpae_iopte pte,
+			       struct io_pgtable_cfg *cfg)
+{
+	smp_mb();
```

Comment:
I still don't know exactly what are the proper barriers before and after writing entry.
Taking into account that:
1. On ARM 64 Xen uses "dsb sy;" before and after writing a pagetable entry
2. On ARM 32 Xen uses "dsb;" before and after writing a pagetable entry
the barrier after mapping should be the same as barrier before mapping and it seems that it is a data synchronization barrier for all operations. So, I think  that we should use mb(); in both cases.
Maybe other guys, clarify this.